src/compiler/scanner.ts

```diff
@@ -2384,15 +2444,979 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
                 }
                 p++;
             }
-
-            while (p < end && isIdentifierPart(text.charCodeAt(p), languageVersion)) {
+            const isUnterminated = !!(tokenFlags & TokenFlags.Unterminated);
+            const endOfBody = p - (isUnterminated ? 0 : 1);
+            let regExpFlags = RegularExpressionFlags.None;
+            while (p < end) {
+                const ch = text.charCodeAt(p);
+                if (!isIdentifierPart(ch, languageVersion)) {
+                    break;
+                }
+                const flag = characterToRegularExpressionFlag(String.fromCharCode(ch));
+                if (flag === undefined) {
+                    error(Diagnostics.Unknown_regular_expression_flag, p, 1);
+                }
+                else if (regExpFlags & flag) {
+                    error(Diagnostics.Duplicate_regular_expression_flag, p, 1);
+                }
+                else if (((regExpFlags | flag) & RegularExpressionFlags.UnicodeMode) === RegularExpressionFlags.UnicodeMode) {
+                    error(Diagnostics.The_Unicode_u_flag_and_the_Unicode_Sets_v_flag_cannot_be_set_simultaneously, p, 1);
+                }
+                else {
+                    regExpFlags |= flag;
+                    const availableFrom = regularExpressionFlagToFirstAvailableLanguageVersion(flag)!;
+                    if (languageVersion < availableFrom) {
+                        error(Diagnostics.This_regular_expression_flag_is_only_available_when_targeting_0_or_later, p, 1, ScriptTarget[availableFrom]);
+                    }
+                }
                 p++;
             }
-            pos = p;
+            pos = tokenStart + 1;
+            const saveTokenPos = tokenStart;
+            const saveTokenFlags = tokenFlags;
+            scanRegularExpressionWorker(text.slice(0, endOfBody), endOfBody, regExpFlags, isUnterminated);

```

Comment:
I did this in order to prevent characters beyond the end of the regex from being accidentally grabbed. `text` without any slicing with adequate `pos <= endOfRegexBody` checks is actually fine.

Reply:
I'd prefer the checks over the extra string slice as that can add up to a fair bit of overhead if you're parsing a bunch of regex literals near the end of a large file. We need to be doing bounds checks anyways, so I'd rather them be against `endOfBody` than `text.length`.